clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -41,6 +42,139 @@ struct is_iterator<T *> : std::true_type {};
+  }
+
+  template <typename T, typename Size, typename Value>
+  static typename ::std::enable_if_t<
+      __has_construct<_Allocator, T *, const Value &>::value, void>
+  uninitialized_fill_n(_Allocator alloc, T *first, Size n, const Value &value) {
+    for (Size i = 0; i < n; i++) {
+      ::std::allocator_traits<_Allocator>::construct(alloc, first + i, value);
+    }
+  }
+
+  template <typename T, typename Size, typename Value>
+  static typename ::std::enable_if_t<
+      !__has_construct<_Allocator, T *, const Value &>::value, void>
+  uninitialized_fill_n(_Allocator alloc, T *first, Size n, const Value &value) {
+    ::std::uninitialized_fill_n(
+        oneapi::dpl::execution::make_device_policy(get_default_queue()), first,
+        n, value);
+  }
+
+  template <typename Iter1, typename Size, typename T>
+  static void uninitialized_custom_copy_n(_Allocator alloc, Iter1 first, Size n,
+                                          T *d_first) {
+    for (Size i = 0; i < n; i++) {
+      ::std::allocator_traits<_Allocator>::construct(alloc, d_first + i,
+                                                     *(first + i));
+    }
+  }
+
+  template <typename Iter1, typename Size, typename T>
+  static typename ::std::enable_if_t<
+      __has_construct<
+          _Allocator, T *,
+          typename ::std::iterator_traits<Iter1>::value_type>::value,
+      void>
+  uninitialized_device_copy_n(_Allocator alloc, Iter1 first, Size n,
+                              T *d_first) {
+    uninitialized_custom_copy_n(alloc, first, n, d_first);
+  }
+
+  template <typename Iter1, typename Size, typename T>
+  static typename ::std::enable_if_t<
+      !__has_construct<
+          _Allocator, T *,
+          typename ::std::iterator_traits<Iter1>::value_type>::value,
+      void>
+  uninitialized_device_copy_n(_Allocator alloc, Iter1 first, Size n,
+                              T *d_first) {
+    ::std::uninitialized_copy_n(
+        oneapi::dpl::execution::make_device_policy(get_default_queue()), first,
+        n, d_first);
+  }
+
+  template <typename Iter1, typename Size, typename T>
+  static typename ::std::enable_if_t<
+      __has_construct<
+          _Allocator, T *,
+          typename ::std::iterator_traits<Iter1>::value_type>::value,
+      void>
+  uninitialized_host_copy_n(_Allocator alloc, Iter1 first, Size n, T *d_first) {
+    uninitialized_custom_copy_n(alloc, first, n, d_first);
+  }
+
+  template <typename Iter1, typename Size, typename T>
+  static typename ::std::enable_if_t<
+      !__has_construct<
+          _Allocator, T *,
+          typename ::std::iterator_traits<Iter1>::value_type>::value,
+      void>
+  uninitialized_host_copy_n(_Allocator alloc, Iter1 first, Size n, T *d_first) {
+    ::std::uninitialized_copy_n(first, n, d_first);
+  }
+
+  template <typename T, typename Size>
+  static
+      typename ::std::enable_if_t<__has_destroy<_Allocator, T *>::value, void>
+      destroy_n(_Allocator alloc, T *p, Size n) {
```

Comment:
Do we want some assert to check that `p != nullptr` throughout these utilities or is that unnecessary here?

Reply:
Good idea, added these asserts.